# Chapter 1: Artificial intelligence

We will start this course introducing the core conceps in which our Artificial Inteligence(AI) theoretical framework relies. 

As AI is about dealing with a system correctly in order to meet some predefined conditions. In the first chapters of this course we will illustrate all the explained concepts using a system called "The taxi system". This is a rather simple system(AI problem) that will allow us to go deep in the fundamental conceps of AI without having to worry about the complex math involved in more sophisticated problems.


## 1.1 What is Artificial intelligence

We will call an AI problem a system, which involves the interactions between two entitites: The agent and the environment. Let's see what they are and how they behave using our Taxi system example.

Starting from scratch, we are confronted to deal with an entity called environment. We know the following things about the environment:
- The agent can observe it.
- The agent can act on it.
- After acting on the environment the observations will change.

In our example we will deal with an environment called the taxi environment. The taxi environment has the following properties:
* When starting the environment we will get a number ranging from 0 to 499.


* There are 6 different possible actions to perform at any given time. We can perform actions by providing the environment with numbers ranging from 0 to 5.




* After performing an action we will get a number ranging from 0 to 499 that we will call observation.
Sometimes after making an action the environment stops working. The end variable will be True in case the environment is over.

<img src="chapter_1_data/environment.png"></img>

Let's see how it works in python:

In [1]:
import gym
sys = gym.make('Taxi-v1') #create a system
sys.reset() #start the environment

[2017-02-08 10:49:50,305] Making new env: Taxi-v1


330

In [2]:
#example of how we can act on an environment
action = 2
observation, _, end, _ = sys.step(action)
observation, end #end will be true if the environment is finished

(350, False)

This means that the only things that we can take as a fact when dealing with a generic system is that the agent will be able to observe it and act on it, knowing that the actions it takes will change its observations. Everything else is what we have to figure out to solve the system.


We would like to know more about our example, so the first thing we will do is playing around with this system. It turns out that it has a nice visual display that allows us to learn a lot from it. We just have to call the render method of our system to get a visual representation of our observations.

In [3]:
sys.step(5)
sys.render()

+---------+
|R: | : :G|
| : : : : |
| : : : : |
| | : | : |
|Y| : |B: |
+---------+
  (Dropoff)


## Things we take for granted about the environment

It turns out that our system represents a videogame following a set of rules:
- It takes place in a 5x5 grid. In one point of the grid there is a square that represents a taxi. We will call each element of the grid point.
- The “|” symbols represent walls. 
- There are 4 points represented by a letter that we will call location. These points are labeled as RBGY.
- There is a passenger in one of the locations.
- The taxi(agent) (represented by a yellow square) can pick up the passenger.
- The agent can drop off the passenger at only one location.
- When the agent drops the passenger the game terminates.
- The pick up and drop off locations are random every time we play the game.
- The starting point of the agent is also random.
- The information about the taxi’s position, the pick up location, the drop off location, and the knowledge that the passenger is alredy inside the taxi is encoded in the observation.


## 1.2 The agent

An agent is an entity of a system that represents the actions you can take on an environment, also called the degrees of freedom of an environment. These interactions are modeled as the possible actions an agent can choose when the system is in a given state, and they will determined by the degrees of freedom an agent has. In our example, the agent that represents the taxi, has only one degree of freedom with 6 different options associated to it.

In order to choose the decision the agent takes at a given state, the agent will obey a set of rules called intelligence. So how we define this intelligence will determine how an agent behaves.

## Things we take for granted about the possible actions
This is what happens when  act on  the environment by providing the following integers:

- **South(0)**: the agent moves Down if there is no wall blocking its way.
- **North(1)**: the agent moves Up if there is no wall blocking its way.
- **East(2)**: the agent moves Right if there is no wall blocking its way.
- **West(3)**: the agent moves left if there is no wall blocking its way.
- **Pickup(4)**: if the agent is next to the passenger(pickup location) then the passenger enters the taxi. In any other case nothing happens.
- **Dropoff(5)**: if the agent is next to the passenger’s destination(drop of location) and the passenger is already inside the taxi, then game terminates. In any other case nothing happens.


Now that we know how our system works, we have decided that we win when the passenger is dropped off. Now we challenge ourselves into figuring out a set of rules that a computer could follow, in order to choose an action after any given observation, so it can win in the fastest way possible. A game will be the decision process that involves following a given policy from a given initial state to the win condition. In this case the states that meet the end condition will be the games in which the environment terminates.


We will call a set of rules a *policy*, and the *policy* that allows to win the game in the least number of actions is called the *optimum policy*. Note that the way that we define this concept of optimality is totally arbitrary.


**An intelligence is an algorithmic process that allows to find the optimum policy an agent has to follow, according to an arbitrary predefined list of conditions**. This is the true core of AI, the only thing that trully matters. Everything else on this field is just theory, being a theory a way of trying to accomplish this task. 

In order to illustrate what an optimum policy for this game looks like, we will load a precalculated optimum policy, and we will use it to play a few games on the taxi system. This way, we will be able to understand better what we are trying to accomplish.

In [4]:
import pickle
from IPython.core.display import clear_output,HTML
import time
with open('chapter_1_data/optimum_policy.pck', 'rb') as handle:
    optimum_policy = pickle.load(handle)
optimum_policy

{'0': 4,
 '1': 4,
 '10': 0,
 '100': 1,
 '101': 1,
 '102': 1,
 '103': 1,
 '104': 2,
 '105': 2,
 '106': 2,
 '107': 2,
 '108': 0,
 '109': 0,
 '11': 0,
 '110': 0,
 '111': 0,
 '112': 0,
 '113': 2,
 '114': 0,
 '115': 2,
 '116': 1,
 '117': 2,
 '118': 0,
 '119': 0,
 '12': 2,
 '120': 3,
 '121': 3,
 '122': 1,
 '123': 3,
 '124': 2,
 '125': 2,
 '126': 2,
 '127': 2,
 '128': 3,
 '129': 3,
 '13': 0,
 '130': 3,
 '131': 3,
 '132': 2,
 '133': 0,
 '135': 2,
 '136': 3,
 '137': 2,
 '138': 3,
 '139': 0,
 '14': 0,
 '140': 3,
 '141': 3,
 '142': 3,
 '143': 3,
 '144': 1,
 '145': 2,
 '146': 2,
 '147': 2,
 '148': 3,
 '149': 3,
 '15': 0,
 '150': 3,
 '151': 0,
 '152': 0,
 '153': 2,
 '155': 2,
 '156': 3,
 '157': 1,
 '158': 0,
 '159': 2,
 '16': 5,
 '160': 3,
 '161': 3,
 '162': 3,
 '163': 3,
 '165': 2,
 '166': 2,
 '167': 1,
 '168': 0,
 '169': 3,
 '17': 0,
 '170': 0,
 '171': 0,
 '172': 0,
 '173': 0,
 '175': 0,
 '176': 3,
 '177': 1,
 '178': 3,
 '179': 0,
 '18': 0,
 '180': 3,
 '181': 3,
 '182': 3,
 '183': 3,
 '185': 1,
 

As we can see, the optimum policy is a one to one correspondence between a possible observation we can get from the environment, and an action chosen from our set of possible actions. With the following function we can use the optimum policy as an "autopilot" for the agent.

In [5]:
def step_taxi(optimum_policy,n_games=10):
    games_solved = 0
    taxi = gym.make('Taxi-v1')
    while games_solved<n_games:
        obs = taxi.reset()
        end = False
        while not end:
            action = optimum_policy[str(obs)]
            obs,_,end,_ = taxi.step(action)
            clear_output(True)
            taxi.render()
            time.sleep(0.25)
        print("Solved game {}".format(games_solved))
        time.sleep(0.75)
        games_solved += 1

In [6]:
step_taxi(optimum_policy,n_games=5)

+---------+
|R: | : :G|
| : : : : |
| : : : : |
| | : | : |
|Y| : |B: |
+---------+
  (Dropoff)
Solved game 4


## 1.3 The state

### 1.3.1 Defining the state

A state is some piece of information that gives us a complete description of a system.

 Let’s explain the concept behind what complete means: If the agent is given any state that corresponds to an environment that you can model, the agent should be able to finish that environment without needing to know anything about the previous states the environment was in.


In our taxi example a suitable definition of state could be the number that our environment spits out every time the agent act on it. The state (represented by a number), can be mapped to a vector that contains the information about the current x and y coordinates of the taxi, the passenger's location, and the passenger's destination.  This four values [x,y,loc,dest] are the minimum information required to be able to play correctly the game. 


Note that if we wanted to play the game without any of that 4 values if would become literally impossible, as those values are specifically mentioned in the rules of the game. For example, we couldn’t win the game if there was no drop off location.



Now let's see an example of what a state is. We can see three different ways of representing the same state: 

- The observation.(We agreed on that this is equivalent to the environment's state)
- The decoded vector.
- The rendered image.

In [7]:
from taxi_model import TaxiModel
tm = TaxiModel()

[2017-02-08 10:50:08,057] Making new env: Taxi-v1


25


We will start defining a measure of how different two states are, that doesn't depend on how we define an optimal behaviour in our system. This number will be the distance. The distance indicates how "far" in the state space are two states are. This is an an arbitrary proprerty that allow us to use some additional infomation we have about our system when comparing two states. 

In this system, after playing some games, it is intuitively easy to tell that a good measure of distance would be the number of actions that the agent has to take to go from one state to the other. We will call this distance the action distance. As this may be something not so easy to figure out in many systems, in the following examples we will compare three different distances.

We will do this in order to see how the choice of a distance represents the knowledge we have about the structure of the system.

The observation is the state representation that gives us the less information about our environment, as it is a single number.

In this case we will define the distance as the difference of the two integers representign two different states.

This definition of distance has (at least) one problem. When getting a number we are only able to tell that two states are different, but we are not even sure that closer numbers for labeling two different states, represent two states that are actually "close" according to the action distance. Let's see what we mean in this example:

In [8]:
stnum_0 = tm.state
tm.env.render()
stnum_0

+---------+
|R: | : :G|
| : : : : |
| : : : : |
| | : | : |
|Y| : |B: |
+---------+



25

We can see that the number 25 corresponds to the displayed position, but if we step the taxi by acting on it just once, we get that the new state is labeled by the number 125. This two states are really close if we think in terms of the number of actions required to get from one state to another(action distance), but if we compare the two numbers using the number distance, we get a value of 100.

In [9]:
tm.step(0)
stnum_1 = tm.state
tm.env.render()
stnum_1

+---------+
|R: | : :G|
| : : : : |
| : : : : |
| | : | : |
|Y| : |B: |
+---------+
  (South)


125

These two numbers 125,25, that are not close in terms of number distance, become really close in terms of aaction distance. This means that we can only infer when we get a 125 that there are at least 124 more states, but we can infer nothig about the real closeness of two states by just comparing the number distance.

Now we move on to the **vector** representation:

A decoded vector follows this structure: [Taxi row, Taxi column, Passenger location, Dropoff location]

This means that as the game takes place in a 5x5 grid, the first two values of the vector will range from 0 to 4. The passenger can be at any of the 4 locations or he can be inside the taxi, so the third number will range from 0 to 4, and the last one will range from 0 to 3, as there are only 4 different dropoff locations.

The vector distance will be defined as the euclidean norm of the difference of the two state vectors.

In [10]:
n_states = 5 * 5 * (4+1) *4
n_states

500

In [11]:
list(tm.decode(125)),list(tm.decode(25))

([1, 1, 1, 1], [0, 1, 1, 1])

The vector representation allows us to have a more structured representation of our states. Here we can see that the euclidean distance between the two vectors would be 1, so at least it looks a bit better than seeing just the number.

Although closer distances tend to correspond to closer states, we should not think of this representation as a silver bullet. It is easy to see that there will be states that are close in vector distance that turn out to be far in terms of "number of actions distance".

In [12]:
from IPython.core.display import clear_output
steps = [0,0,0,0,2]
for s in steps:
    tm.step(s)
    clear_output()
tm.env.render()
tm.state

+---------+
|R: | : :G|
| : : : : |
| : : : : |
| | : | : |
|Y| : |B: |
+---------+
  (East)


445

In [13]:
state_0 = list(tm.decode(tm.state))

In [14]:
steps = [1,1,2,0,0] #it takes 5 steps to get from one place to the other
for s in steps:
    tm.step(s)
    clear_output()
tm.env.render()
state_1 = list(tm.decode(tm.state))
tm.state,state_1

+---------+
|R: | : :G|
| : : : : |
| : : : : |
| | : | : |
|Y| : |B: |
+---------+
  (South)


(465, [4, 3, 1, 1])

In [15]:
import numpy as np
print("Num distance:{}, euclidean distance: {}, action distance:{}".format(465-445,
                                                                            np.linalg.norm(np.array(state_0)\
                                                                                        -np.array(state_1)),
                                                                            len(steps)))

Num distance:20, euclidean distance: 1.0, action distance:5


This happens because there is no information about the walls in the grid contained in the vector representation. If instead of using the vector representation we use the rendered image of our environment it is easy to calculate the action distance between two different states. This means that although a state represents a complete description of a system, it could only contains the minimum information needed to distinguish between two different states, and not the information needed to know for going from one place to the other.

You could however, infer how one state is connected to another by repeatedly observing and acting on the environment. Sadly this is what modeling is about, so you will have to wait a bit to know about it. From now on we will assume that we have a perfect model of our state space.



### 1.3.2 The state space

The state space, also called phase space in physics, is a methematical entity that includes the information about the states of the system, and how that states are connected to each other. This is how all the different states come toghether to define our system.

The state space not only contains all the different states but also the possible transitions between states. A transition associated between two states is how you have to act on your environment to get from one state to the other.

It is easy to analyse the structure of the state space by using graphs. 

<img src="chapter_1_data/graph.png" width="200px" height="150px">

*In mathematics graph theory is the study of graphs, which are mathematical structures used to model pairwise relations between objects. A graph in this context is made up of vertices, nodes, or points which are connected by edges, arcs, or lines* <a href="https://en.wikipedia.org/wiki/Graph_theory">More on graph theory</a>

In our context we will be using graphs to represent our state space, by assigning one state to each node. Then we will add an edge between two nodes A and B if there is one action that allows to go from state A to state B. 

If we do this we will end up with a network that gives us a complete description of our system. We will load the graph representation of our state space to see if we can gain some insight on how the taxi game works.

First we load the plots:

In [16]:
import pickle
from bokeh.io import output_notebook
output_notebook(hide_banner=True)
with open('chapter_1_data/state_space_general.pck', 'rb') as handle:
    overview = pickle.load(handle)
with open('chapter_1_data/state_space_zoom_transition.pck', 'rb') as handle:
    transitions = pickle.load(handle)
with open('chapter_1_data/state_space_vector_info.pck', 'rb') as handle:
    vectors = pickle.load(handle)
with open('chapter_1_data/state_space_opt_policy.pck', 'rb') as handle:
    graph_policy = pickle.load(handle)
with open('chapter_1_data/optimum_policy_tree.pck', 'rb') as handle:
    tree_policy = pickle.load(handle)

This is a little hack so we can see larger plots:

In [17]:
%%HTML
<style>
.container{width:90% !important;}
</style>

Now let's take a look at the following visual representations of the state space. We have calculated the graph of the environment, and we have represented different aspects the information we have so far about the taxi system. Here we can see that each node represents a state. The blue ones are the states in wich the game can end, and the black ones are all the other states. We also have asigned each edge a color depending on the action that has to be chosen in order to get from one state to the other.

If we take a look at the structure of the state space we can see that there are four unconnected components, each one with a blue state in it. This is because all the states in every one of the four components have the same dropoff location.

Now please zoom into one of its four regions using the zoom tool of the chart.

You can also see a square box in the center of each edge. This box es colored by the action needed to go from one place to another. Note that when you hover the mouse over that box it displays the information represented on the edge. Between two states A and B there are two links, one that represents an action that takes you from A to B, and another one that represents the action required to go from B to A. This is something that may not happen in every problem, as there could be states that are only reachable in one direction. If that is the case, we will call this fenomena irreversibility.

In [18]:
from IPython.core.display import HTML
HTML(overview)

#### Visualizing how actions link different states

Graphs are a useful tool to get a grasp on what is going on in our state space. In the next plot you can see how all the different states are related to each other by the actions that connects them. We have emphatised the connections between states in order to see how the actions taken define a path through the state space graph. Every sequence of actions an agent takes can be seen as following a path in the graph. A game will finish once the agent follows a path connecting its initial state with the end state. (Blue state in the former plot)

In every of the 4 components we can see a similar structure:

- 5 different regions separated by the pickup and dropoff actions.
- The central region is accessible once the passenger is picked up. 
- A game is always started in one of the four outer branches of each game.


In [19]:
HTML(transitions)

Note how the numbers of the different states are related to each other, as this generic view offers new insight on how the states were originally presented to us, we can see how the number distance relates to the action distance.

#### Visualizing vector information

 We can also represent information conatined in our vector represenation to gain insight on the structure of the state space.

In [20]:
HTML(vectors)

In this figure we have colored every node by its corresponding passenger's location. Now it is easier to see how the five regions of each component are connected to each other. Each time we play a different game we will be randomly placed in one state (node) located in one of the four outer regions of any of the four possible components. This means that there will be 16 possible combinations of pickup/dropoff locations. 

#### Representing a policy on the state space graph

Taking into account the objective of our game, if we use our state space representation, our problem definition is analogous to finding the shortest path between the initial state and the end state of each game. As every action represents a link between two states, once we have found the path in our graph, we just have to take the actions that correspond to each of the path's edges.

It is also possible to represent the optimum policy in the state space graph. In the next figure we will superimpose the optimum policy over all the state space:

In [21]:
HTML(graph_policy)

In this plot you can see how the optimum policy represents a one-way path that links almost all the states in a tree like fashion, always ending up in the end condition. If you are able to find this optimum policy tree we will consider the system solved. we know that this is nothing more than restating what we said at the begining of the lesson, but if we watch carefully our graph we can see that there are some states that have not any decision associated with them.

That states that are "out of the tree" represent states that will never be reachable if an agent follows the optimum policy, making its contribution totally irrelevant to the solution of the system. This is not a minor thing, as this means that it is possible to find an optimum policy for an environment without having to explicitly evaluate all the possible states an agent could be in. If our model had absolute no information about the states that are left out the optimum policy we would still be able to solve the problem.

In this 500 states problem this may not seem important, but remember that this is only an easy example to get an intuitive approach on how AI works. In more sophisticated problems our graph will literally have infinite nodes, so knowing that we don't need to evaluate all the possible states to solve a problem is a really good thing. This means that we could be able to find the optimum policy without evaluating all the states, but we wouldn't be able to prove that the policy found is optimum.

The following graph is just another way of representing the optimum policy tree. As you can see there are 4 trees, each one corresponding to a component with a different dropoff location. We have colored the nodes by its passenger location, and the edges by the actions needed to go from one state to the other. 

In [22]:
HTML(tree_policy)

If you wanted to use the plot above for solving any taxi game, you would just have to find your initial state in the tree, and follow the transitions downwards by choosing the action corresponding to the color of each edge you pass through.